File: frontend/src/api/attendance.ts
import apiClient from './client'
import type { ApiResponse, PageResult } from './types'

// 重新导出共享类型，便于视图层从 attendance.ts 统一导入
export type { PageResult }

export interface ShiftConfig {
  id: string
  shift_code: string
  shift_name: string
  start_time: string
  end_time: string
  is_overnight: boolean
  color: string
  sort_order: number
  is_active: boolean
}

export interface ScheduleCell {
  date: string
  scheduled_shift: string | null
  shift_start_time: string | null
  shift_end_time: string | null
  is_overnight: boolean
  clock_in: string | null
  clock_out: string | null
  status: string
  late_minutes: number
  early_minutes: number
}

export interface ScheduleTableData {
  date_from: string
  date_to: string
  days: string[]
  employees: {
    employee_id: string
    employee_name: string
    employee_role: string
    cells: ScheduleCell[]
  }[]
  shifts: {
    shift_code: string
    shift_name: string
    color: string
  }[]
  stats: Record<string, { work: number; rest: number; late: number; absent: number }>
}

export interface MyScheduleData {
  employee_id: string
  date_from: string
  date_to: string
  cells: ScheduleCell[]
  shifts: {
    shift_code: string
    shift_name: string
    color: string
  }[]
  stats: {
    work_days: number
    present_days: number
    late_days: number
    leave_days: number
    absent_days: number
  }
}

/** 排班批量保存结果 */
export interface ScheduleBatchResult {
  created: number
  updated: number
}

/** 智能排班生成结果 */
export interface ScheduleGenerateResult {
  schedules: Array<{
    employee_id: string
    employee_name: string
    date: string
    scheduled_shift: string
  }>
  stats: {
    total_entries: number
    employees: number
    rest_days: number
  }
}

/** 员工排班规则更新结果 */
export interface EmployeeRuleUpdateResult {
  updated: number
}

/** 打卡记录 */
export interface AttendanceRecord {
  id: string
  employee_id: string
  store_id: string
  date: string
  scheduled_shift: string | null
  clock_in: string | null
  clock_out: string | null
  status: string
  late_minutes: number
  early_minutes: number
  source: string
  note: string | null
  employee_name: string
  deduction: number
  deduction_reason: string
}

/** 今日打卡状态条目（复用 AttendanceRecord 结构） */
export type TodayAttendanceItem = AttendanceRecord

/** 指纹机同步结果 */
export interface FingerprintSyncResult {
  synced_count: number
  new_records: number
  updated_records: number
  errors: string[]
}

/** 补卡申请参数 */
export interface MakeupRequestParams {
  employee_id: string
  date: string
  clock_in: string | null
  clock_out: string | null
  reason: string
}

/** 补卡申请结果 */
export interface MakeupResult {
  id: string
  fee: number
  month_count: number
}

/** 请假申请参数 */
export interface LeaveRequestParams {
  employee_id: string
  start_date: string
  end_date: string
  leave_type: string
  reason: string
}

/** 请假记录 */
export interface LeaveResponse {
  id: string
  employee_id: string
  employee_name: string
  start_date: string
  end_date: string
  leave_type: string
  reason: string
  status: string
  created_at: string | null
}

/** 请假记录列表响应 */
export interface LeaveRecordsData {
  items: LeaveResponse[]
  total: number
  page: number
  page_size: number
  total_pages: number
}

/** 月末汇总条目 */
export interface MonthlySummaryItem {
  employee_id: string
  employee_name: string
  employee_role: string
  total_days: number
  present_days: number
  late_days: number
  early_days: number
  leave_days: number
  absent_days: number
  makeup_count: number
  late_fee: number
  makeup_fee: number
  total_deduction: number
}

/** 月末汇总响应 */
export interface MonthlySummaryData {
  period: string
  store_id: string
  items: MonthlySummaryItem[]
  total_deduction: number
}

/** 打卡记录查询参数 */
export interface RecordsQueryParams {
  employee_id?: string
  date_from?: string
  date_to?: string
  status?: string
  page?: number
  page_size?: number
}

export const attendanceAPI = {
  /** 获取全员排班考勤表（管理端） */
  getScheduleTable(params: { date_from: string; date_to: string }) {
    return apiClient.get<ApiResponse<ScheduleTableData>>('/attendance/schedule', { params })
  },

  /** 获取个人排班考勤（员工端） */
  getMySchedule(params: { date_from: string; date_to: string }) {
    return apiClient.get<ApiResponse<MyScheduleData>>('/attendance/schedule/my', { params })
  },

  /** 批量保存排班 */
  batchSaveSchedules(data: { schedules: Array<{ employee_id: string; date: string; scheduled_shift: string }> }) {
    return apiClient.post<ApiResponse<ScheduleBatchResult>>('/attendance/schedule/batch', data)
  },

  /** 智能排班生成 */
  generateSchedule(data: { year: number; month: number }) {
    return apiClient.post<ApiResponse<ScheduleGenerateResult>>('/attendance/schedule/generate', data)
  },

  /** 获取班次配置列表 */
  getShiftConfigs() {
    return apiClient.get<ApiResponse<ShiftConfig[]>>('/attendance/shifts')
  },

  /** 保存班次配置 */
  saveShiftConfig(data: Partial<ShiftConfig> & { shift_code: string; shift_name: string; start_time: string; end_time: string }) {
    return apiClient.post<ApiResponse<ShiftConfig>>('/attendance/shifts', data)
  },

  /** 同步企微打卡 */
  syncCheckin(data?: { target_date?: string }) {
    return apiClient.post<ApiResponse<{ synced_count: number; errors: string[] }>>('/attendance/sync', data || {})
  },

  /** 批量更新员工排班规则 */
  batchUpdateEmployeeRules(data: {
    employees: Array<{
      employee_id: string
      shift_group: string | null
      is_first_manager: boolean
      is_second_manager: boolean
      is_third_manager: boolean
    }>
  }) {
    return apiClient.put<ApiResponse<EmployeeRuleUpdateResult>>('/attendance/employee-rules', data)
  },

  /** 获取今日打卡状态 */
  getToday() {
    return apiClient.get<ApiResponse<TodayAttendanceItem[]>>('/attendance/today')
  },

  /** 同步指纹机打卡 */
  syncFingerprint(date: string) {
    return apiClient.post<ApiResponse<FingerprintSyncResult>>('/attendance/sync/fingerprint', { target_date: date })
  },

  /** 申请补卡 */
  requestMakeup(data: MakeupRequestParams) {
    return apiClient.post<ApiResponse<MakeupResult>>('/attendance/makeup', data)
  },

  /** 申请请假 */
  requestLeave(data: LeaveRequestParams) {
    return apiClient.post<ApiResponse<LeaveResponse>>('/attendance/leave', data)
  },

  /** 获取请假记录列表 */
  getLeaveRecords(params: { status?: string; page?: number; page_size?: number } = {}) {
    return apiClient.get<ApiResponse<LeaveRecordsData>>('/attendance/leave', { params })
  },

  /** 审批请假 */
  approveLeave(recordId: string, data: { action: string; resolution?: string }) {
    return apiClient.put<ApiResponse<LeaveResponse>>(`/attendance/leave/${recordId}/approve`, data)
  },

  /** 获取月度考勤汇总 */
  getMonthlySummary(period: string) {
    return apiClient.get<ApiResponse<MonthlySummaryData>>('/attendance/monthly', {
      params: { period },
    })
  },

  /** 获取打卡记录列表 */
  getRecords(params: RecordsQueryParams) {
    return apiClient.get<ApiResponse<PageResult<AttendanceRecord>>>('/attendance/records', { params })
  },
}

File: backend/app/schemas/attendance.py
"""
考勤与排班一体化 Pydantic 模型
"""
import uuid
from datetime import date
from pydantic import BaseModel, Field, field_validator


# ==================== 排班表 ====================

class ScheduleCell(BaseModel):
    date: str
    scheduled_shift: str | None = None
    status: str = "unknown"
    clock_in: str | None = None
    clock_out: str | None = None
    late_minutes: int = 0
    early_minutes: int = 0


class ScheduleEmployeeRow(BaseModel):
    employee_id: uuid.UUID
    employee_name: str
    employee_role: str
    cells: list[ScheduleCell]


class ScheduleShiftInfo(BaseModel):
    shift_code: str
    shift_name: str
    color: str


class ScheduleTableResponse(BaseModel):
    date_from: str
    date_to: str
    shifts: list[ScheduleShiftInfo]
    employees: list[ScheduleEmployeeRow]


class ScheduleBatchItem(BaseModel):
    employee_id: uuid.UUID
    date: str
    scheduled_shift: str


class ScheduleBatchRequest(BaseModel):
    schedules: list[ScheduleBatchItem]


class ScheduleBatchResponse(BaseModel):
    created: int
    updated: int


class MyScheduleResponse(BaseModel):
    employee_id: uuid.UUID
    date_from: str
    date_to: str
    shifts: list[ScheduleShiftInfo]
    cells: list[ScheduleCell]
    stats: dict


# ==================== 班次配置 ====================

class ShiftConfigItem(BaseModel):
    id: uuid.UUID | None = None
    shift_code: str
    shift_name: str
    start_time: str
    end_time: str
    is_overnight: bool = False
    color: str = "#FB0079"
    sort_order: int = 0
    is_active: bool = True

    @field_validator("start_time", "end_time")
    @classmethod
    def validate_time(cls, v: str) -> str:
        import re
        if not re.match(r"^\d{2}:\d{2}$", v):
            raise ValueError("时间格式必须为 HH:MM")
        return v


class ShiftConfigSaveRequest(BaseModel):
    shift_code: str
    shift_name: str
    start_time: str
    end_time: str
    is_overnight: bool = False
    color: str = "#FB0079"
    sort_order: int = 0
    is_active: bool = True


# ==================== 打卡记录 ====================

class AttendanceRecordResponse(BaseModel):
    id: uuid.UUID
    employee_id: uuid.UUID
    store_id: uuid.UUID
    date: str
    scheduled_shift: str | None = None
    clock_in: str | None = None
    clock_out: str | None = None
    status: str
    late_minutes: int = 0
    early_minutes: int = 0
    source: str = "manual"
    note: str | None = None
    employee_name: str = ""
    deduction: float = 0.0
    deduction_reason: str = ""


# ==================== 智能排班生成 ====================

class ScheduleGenerateRequest(BaseModel):
    year: int
    month: int


class ScheduleGenerateItem(BaseModel):
    employee_id: uuid.UUID
    employee_name: str = ""
    date: str
    scheduled_shift: str


class ScheduleGenerateResponse(BaseModel):
    schedules: list[ScheduleGenerateItem]
    stats: dict


class EmployeeRuleItem(BaseModel):
    employee_id: uuid.UUID
    shift_group: str | None = None
    is_first_manager: bool = False
    is_second_manager: bool = False
    is_third_manager: bool = False


class EmployeeRuleUpdateRequest(BaseModel):
    employees: list[EmployeeRuleItem]

    model_config = {"from_attributes": True}


class AttendanceRecordCreate(BaseModel):
    employee_id: uuid.UUID
    date: str
    clock_in: str | None = None
    clock_out: str | None = None
    scheduled_shift: str | None = None


class AttendanceListRequest(BaseModel):
    employee_id: uuid.UUID | None = None
    date_from: str | None = None
    date_to: str | None = None
    status: str | None = None
    page: int = 1
    page_size: int = 20


# ==================== 同步 ====================

class SyncRequest(BaseModel):
    target_date: str | None = None


class SyncResponse(BaseModel):
    synced_count: int
    errors: list[str]


# ==================== 月末汇总 ====================

class MonthlySummaryItem(BaseModel):
    employee_id: uuid.UUID
    employee_name: str
    employee_role: str
    total_days: int = 0
    present_days: int = 0
    late_days: int = 0
    early_days: int = 0
    leave_days: int = 0
    absent_days: int = 0
    makeup_count: int = 0
    total_late_minutes: int = 0


class MonthlySummaryResponse(BaseModel):
    period: str
    store_id: uuid.UUID
    items: list[MonthlySummaryItem]


# ==================== WiFi打卡绑定 ====================

class CheckinWifiItem(BaseModel):
    id: uuid.UUID | None = None
    ssid: str
    bssid: str
    label: str | None = None
    is_active: bool = True

    model_config = {"from_attributes": True}


class CheckinWifiCreateRequest(BaseModel):
    ssid: str = Field(..., min_length=1, max_length=64)
    bssid: str = Field(..., min_length=1, max_length=32)
    label: str | None = None
    is_active: bool = True


class CheckinConfigResponse(BaseModel):
    """打卡配置（前端用于决定打卡流程）"""
    require_wifi: bool = True
    require_photo: bool = True
    grace_minutes: int = 5
    photo_retention_days: int = 90
    time_window_minutes: int = 120
    wifis: list[CheckinWifiItem] = []


class CheckinConfigUpdateRequest(BaseModel):
    require_wifi: bool | None = None
    require_photo: bool | None = None
    grace_minutes: int | None = Field(None, ge=0, le=120)
    photo_retention_days: int | None = Field(None, ge=7, le=365)
    time_window_minutes: int | None = Field(None, ge=30, le=360)


# ==================== 打卡结果 ====================

class CheckinResultResponse(BaseModel):
    """打卡结果"""
    record_id: uuid.UUID
    date: str
    action: str  # clock_in / clock_out
    clock_time: str
    scheduled_shift: str | None = None
    shift_start_time: str | None = None
    shift_end_time: str | None = None
    status: str  # present / late / early / unknown
    late_minutes: int = 0
    early_minutes: int = 0
    matched_by: str  # schedule / time_window
    photo_url: str | None = None
    wifi_ssid: str | None = None


class CheckinStatusResponse(BaseModel):
    """今日打卡状态（前端打卡页展示用）"""
    date: str
    scheduled_shift: str | None = None
    shift_start_time: str | None = None
    shift_end_time: str | None = None
    is_overnight: bool = False
    clock_in: str | None = None
    clock_out: str | None = None
    status: str = "unknown"
    late_minutes: int = 0
    early_minutes: int = 0
    can_checkin: bool = True
    next_action: str = "clock_in"  # clock_in / clock_out / done
